### PR TITLE
Added settings with test cases.

### LINKED ISSUE #10: feat(unti-test): introduce
# Context

We have to test our app, it's very important to provide the best user experience, and avoid bugs. But in the javascript ecosystem, there are plenty of ways to test your project, so you need to choose the right one.

here's a list of tools:
- [jest](https://nextjs.org/docs/app/building-your-application/testing/jest)
- [Playwright](https://nextjs.org/docs/app/building-your-application/testing/playwright)
- [cypress](https://nextjs.org/docs/app/building-your-application/testing/cypress)
- [node.js internal test-runner](https://nodejs.org/docs/latest/api/test.html)

> If you have any other tools you like tell it

There are still things to be decided. What tools to use, what to test?

# Todo

- [ ] reached a consensus on how and what to test.
- [ ] added test.
- [ ] document how to test different parts of the code.

### LINKED ISSUE #14: feat(backend) add settings loading mechanism

### PR DESCRIPTION
**Pull request by SamTV12345**
## Description

Add back-end logic and small test.

## Related Issues

For test #10 
For back-end #14 

### Check List

- [x] I have formatted using `npx turbo format`
- [x] I have checked the types using `npx turbo type-check`
- [x] I have tested the build using `npx turbo build`

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -44,3 +44,5 @@ next-env.d.ts
 
 # JetBrains
 .idea
+
+/settings.json
```
Comment (AugustinMauroy) [inline .gitignore]:
```suggestion

# Production configuration file
/settings.json
```

File: backend/utils/AbsolutePaths.ts [hunk 2]
```diff
@@ -0,0 +1,34 @@
+import path from 'path';
```
Comment (AugustinMauroy) [inline backend/utils/AbsolutePaths.ts]:
```suggestion
import path from 'node:path';
```

File: tests/settings/settings.spec.test.ts [hunk 3]
```diff
@@ -0,0 +1,70 @@
+'use strict';
+
+const assert = require('assert').strict;
```
Comment (AugustinMauroy) [inline tests/settings/settings.spec.test.ts]:
```suggestion
import assert from 'node:assert/strict';
```

[node.js docs](https://nodejs.org/docs/latest/api/assert.html#strict-assertion-mode)

File: backend/utils/minify.ts [hunk 4]
```diff
@@ -0,0 +1,76 @@
+const minify = (json: string) => {
```
Comment (AugustinMauroy) [inline backend/utils/minify.ts]: what does this code do?

File: tests/settings/settings.spec.test.ts [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+import settingsMod from '../../backend/utils/Setting';
+import path from 'path';
+import process from 'process';
+import { describe, it, beforeEach } from 'vitest';
+
```
Comment (AugustinMauroy) [inline tests/settings/settings.spec.test.ts:6]: what do you thinks about just Ising nodejs test-runner ?
with the `tsx` tool, which is already in the infra, we can run the test in typescript and node is our runner, so it's officially present. This means less dependency.

in `package.json` add these value:
```json
"test:unit": "cross-env NODE_NO_WARNINGS=1 node --import=tsx --test **/*.test.mjs **/*.test.jsx",
"test:watch": "cross-env NODE_NO_WARNINGS=1 node --import=tsx --test --watch **/*.test.mjs **/*.test.jsx",
```

File: backend/utils/Setting.ts [hunk 6]
```diff
@@ -0,0 +1,786 @@
+'use strict';
```
Comment (AugustinMauroy) [inline backend/utils/Setting.ts]: why we have this header ?

**Review by AugustinMauroy — COMMENTED** [review]
Good starting but some point.

- `backend/utils` -> should be `utils/backend`. For example if it's just string manipulation we will put under `utils/` because this js didn't depend of runtime but if it's client it's will under `utils/client`
- we should use [`pinojs`](https://getpino.io/#/) instead of `log4js` it's super perf. and maintained by Matteo collins (nodejs TSC)
- don't forget to ping the packages for security `~`.

File: backend/utils/AbsolutePaths.ts (hunk 2 above)
Comment (SamTV12345) [reply backend/utils/AbsolutePaths.ts]: What is the difference here? Or is this just for documentation that we use the node path package?

Comment (AugustinMauroy) [reply backend/utils/AbsolutePaths.ts]: I think it's good practice to use the `node:` to differentiate between npm packages and nodejs internal modules.

File: backend/utils/AbsolutePaths.ts [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+const absPathLogger = log4js.getLogger('AbsolutePaths');
+
+export const findEtherpadRoot = () => {
+  if (etherpadRoot != null) {
+    return etherpadRoot;
+  }
+
+  const foundRoot = findRoot(__dirname);
+
+  return foundRoot;
```
Comment (AugustinMauroy) [inline backend/utils/AbsolutePaths.ts]:
```suggestion
  if (etherpadRoot != null) {
    return etherpadRoot;
  }

  return process.cwd();
```

Here we can simplify by using this it's didn't need any packages and it's work same on our infra (I had test on my machine)

File: tests/settings/settings.spec.test.ts (hunk 5 above)
Comment (SamTV12345) [reply tests/settings/settings.spec.test.ts:6]: I guess we need vitest anyway because we also want to write tests for our React components.

Comment (AugustinMauroy) [reply tests/settings/settings.spec.test.ts:6]: with nodejs runner normally it's possible using jsdom.
it's also remove package `vite-tsconfig-paths`

File: backend/utils/AbsolutePaths.ts (hunk 7 above)
Comment (SamTV12345) [reply backend/utils/AbsolutePaths.ts]: True I changed it.

File: backend/utils/Setting.ts (hunk 6 above)
Comment (SamTV12345) [reply backend/utils/Setting.ts]: Sorry was a leftover from the Etherpad repository.

File: backend/utils/AbsolutePaths.ts (hunk 7 above)
Comment (AugustinMauroy) [reply backend/utils/AbsolutePaths.ts]: Perfect ✨

File: tests/settings/settings.spec.test.ts (hunk 5 above)
Comment (SamTV12345) [reply tests/settings/settings.spec.test.ts:6]: Which one is faster?

File: backend/utils/Setting.ts (hunk 6 above)
Comment (AugustinMauroy) [reply backend/utils/Setting.ts]: I suspected as much. That's what the review is for, to avoid keeping legacy code.

Comment (SamTV12345) [reply backend/utils/Setting.ts]: Yeah I'm sometimes a bit blindfolded. So good catch :)

File: backend/utils/minify.ts (hunk 4 above)
Comment (SamTV12345) [reply backend/utils/minify.ts]: It minifies the settings.json and removes all the comments that JSON.parse would not understand.

File: tests/settings/settings.spec.test.ts (hunk 5 above)
Comment (AugustinMauroy) [reply tests/settings/settings.spec.test.ts:6]: when I had work on nodejs.org, I can say nodejs is faster that jest. I'm not expert with vite. But What I know is nodejs is built-in and use maximum of cpu concurrently.

File: backend/utils/minify.ts (hunk 4 above)
Comment (AugustinMauroy) [reply backend/utils/minify.ts]: add just small JSDoc on top like

```js
/*
  * It's allow us to remove comment on json
  */
```

So on ide when we hover function we know what she doe

File: tests/settings/settings.spec.test.ts (hunk 5 above)
Comment (SamTV12345) [reply tests/settings/settings.spec.test.ts:6]: Can it also handle the path aliases?

Comment (AugustinMauroy) [reply tests/settings/settings.spec.test.ts:6]: YES it's handle by `tsx` like in `server.ts`

**Pull request merged**